File: DSA_python/DSA01006 HOÁN VỊ NGƯỢC.py
def p(n):
    total = 1
    for i in range(n):
        total *= (i+1)
    return int(total)

def sort_list(A, start, end):
    B = []
    for i in range(start, end):
        B.append(A[i])
    B.sort(reverse = True)

    j = 0
    for i in range(start, end):
        A[i] = B[j]
        j += 1
    return A

def update(A, n):
    i = n-2
    while A[i] < A[i+1]:
        i -= 1
    if i < 0:
        return
    
    k = n-1
    while A[i] < A[k]:
        k -= 1
    A[i], A[k] = A[k], A[i]
    sort_list(A, i+1, n)
    return A

def submit():
    n = int(input())
    A = []
    for i in range(n):
        A.append(n-i)

    for i in range(p(n)):
        for j in A:
            print(j, end = "")
        print(end = " ")
        update(A, n)

    print()

test_num = int(input())
for i in range(test_num):
    submit()

File: DSA/DSA01006 HOÁN VỊ NGƯỢC.cpp
#include <bits/stdc++.h>
using namespace std;

void update(vector<int> &A, int n)
{
    int i = n-2;
    while (A[i] < A[i+1])
    {
        i--;
    }

    if (i < 0)
    {
        return;
    }
    
    int k = n-1;
    while (A[i] < A[k])
    {
        k--;
    }
    
    swap(A[i], A[k]);
    sort(A.begin()+i+1, A.end(), greater<int>());
}

long long p(int n)
{
    long long total = 1;
    for (int i = 1; i <= n; i++)
    {
        total *= i;
    }
    return total;
}

void submit()
{
    int n;
    cin >> n;
    vector<int> A(n);
    for (int i = 0; i < n; i++)
    {
        A[i] = n-i;
    }
    
    int total = p(n);
    for (int i = 0; i < total; i++)
    {
        for (auto &&j : A)
        {
            cout << j;
        }
        cout << " ";
        update(A, n);
    }
    
    cout << endl;
}   

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();

    while (so_bo_test--)
    {
        submit();
    }

    return 0;
}